Запрос на менторство BJS2-76933

VladimirEkgardt opened the pull request.

VladimirEkgardt committed:
Запрос на менторство BJS2-76933

SlavaVermillion commented on src/main/java/school/faang/user_service/service/MentorshipRequestService.java:
```diff
@@ -0,0 +1,106 @@
+                    requester.getSentMentorshipRequests().add(mapper.toEntity(request));
+                    User receiver = userRepo.findById(request.getReceiverId()).get();
+                    receiver.getReceivedMentorshipRequests().add(mapper.toEntity(request));
+
```
Смотри, в задаче написано, что запрос на менторство можно делать лишь раз в 3 месяца, т. е. если прошло меньше 3 месяцев, то нужно выбросить ошибку, у тебя же здесь, как я понимаю, создается новый запрос

SlavaVermillion commented on src/main/java/school/faang/user_service/service/MentorshipRequestService.java:
```diff
@@ -0,0 +1,106 @@
+        if (!Objects.equals(request.getRequesterId(), request.getReceiverId())) {
+            if (userRepo.findById(request.getRequesterId()).isEmpty() || userRepo.findById(request.getReceiverId()).isEmpty()) {
+                throw new IllegalArgumentException("The user does not exist");
+            }
```
Вот эта конструкция довольно пугающая и не совсем эффективная. По условию нужно проверить, что пользователь, который запрашивает менторство и пользователь, у которого запрашивают - существуют в базе данных, а у тебя получается, что ошибка выбрасывается общая. Я думаю, что лучше сделать так, чтобы можно было узнать, кого конкретно нет в БД и выкидывать об этом ошибку. Сделать можно что-то типо:
requester = userRepo.findById(request.getRequesterId()).orElseThrow()...
receiver = userRepo.findById(request.getReceiverId()).orElseThrow()...

SlavaVermillion commented on src/main/java/school/faang/user_service/service/MentorshipRequestService.java:
```diff
@@ -0,0 +1,106 @@
+
+    public MentorshipRequestDto requestMentorship(MentorshipRequestDto request) {
+        if (!Objects.equals(request.getRequesterId(), request.getReceiverId())) {
+            if (userRepo.findById(request.getRequesterId()).isEmpty() || userRepo.findById(request.getReceiverId()).isEmpty()) {
```
Если честно, это огромная конструкция из if-ов пугает. Может попробуем сделать проще?
Вот это  if (!Objects.equals(request.getRequesterId(), request.getReceiverId())) переделаем вот в это 
if (request.getRequesterId().equals(request.getReceiverId())) {
а тут сразу будем выбрасывать исключение с сообщением, что у самого себя нельзя просить запрос на менторство
}

SlavaVermillion commented on src/main/java/school/faang/user_service/service/MentorshipRequestService.java:
```diff
@@ -0,0 +1,106 @@
+            return request;
+        } else {
+            throw new IllegalArgumentException("The user cannot send a request to himself");
+        }
```
Если мы избавились от всей этих вложенных if, значит 52-54 строчки на не нужны

SlavaVermillion commented on src/main/java/school/faang/user_service/service/MentorshipRequestService.java:
```diff
@@ -0,0 +1,106 @@
+                }
+            } else {
+                mentorshipRequestRepository.create(request.getRequesterId(), request.getReceiverId(), request.getDescription());
+            }
```
Раз мы избавились от внешних и вложенных if-ов, теперь 49-50 строчки не должны быть внутри чего-либо

SlavaVermillion commented on src/main/java/school/faang/user_service/controller/MentorshipRequestController.java:
```diff
@@ -0,0 +1,46 @@
+    @GetMapping
+    public List<MentorshipRequestDto> getRequests(@RequestBody RequestFilterDto filter) {
+        return service.getRequest(filter);
+    }
```
Мелочь, но все же, по задаче метод у MentorshipRequestService должен называться getRequests

SlavaVermillion replied:
Еще не хватает фильтра по описанию. Помню, что мы обсуждали, что это звучит странно, но все же по задаче он должен быть)))

SlavaVermillion commented:
Еще не хватает фильтра по описанию. Помню, что мы обсуждали, что это звучит странно, но все же по задаче он должен быть)))

SlavaVermillion commented:
Реализация фильтров на высоте! Получилось масштабируемо!

SlavaVermillion commented on src/main/java/school/faang/user_service/service/MentorshipRequestService.java:
```diff
@@ -0,0 +1,106 @@
+            User receiver = userRepo.findById(request.getReceiver().getId()).get();
+
+            if (!receiver.getMentees().contains(receiver)) {
+                request.setStatus(RequestStatus.ACCEPTED);//обновяться ли таким образом данные в бд
```
вот тут if (!receiver.getMentees().contains(receiver)) ты проверяешь, что ментор является своим собственным ментором, а по задаче нужно:  Если пользователь уже является ментором (rec) отправителя (req), выбросить исключение с сообщением об этом.
И еще давай request.setStatus(RequestStatus.ACCEPTED); перенесем в самый конец метода

SlavaVermillion commented on src/main/java/school/faang/user_service/service/MentorshipRequestService.java:
```diff
@@ -0,0 +1,106 @@
+                request.setStatus(RequestStatus.ACCEPTED);//обновяться ли таким образом данные в бд
+                receiver.getMentees().add(receiver);
+
+                requester.getMentors().add(requester);
```
Смотри, вот тут у тебя получается, что ученик стал своим же учеником, а ментор стал своим же ментором. Но ведь должно быть:  у ученика появился ментор, у ментора появился ученик

SlavaVermillion commented on src/main/java/school/faang/user_service/service/MentorshipRequestService.java:
```diff
@@ -0,0 +1,106 @@
+            MentorshipRequest request = requestOptionalRequest.get();
+            User requester = userRepo.findById(request.getRequester().getId()).get();
+            User receiver = userRepo.findById(request.getReceiver().getId()).get();
+
```
давай вынесем requester и receiver из if, уберем .get() и сразу накинем на каждый .orElseThrow() с исключением, тогда будет смотреться лучше

SlavaVermillion commented on src/main/java/school/faang/user_service/service/MentorshipRequestService.java:
```diff
@@ -0,0 +1,106 @@
+    public void acceptRequest(Long id) {
+        Optional<MentorshipRequest> requestOptionalRequest = mentorshipRequestRepository.findById(id);
+        if (requestOptionalRequest.isPresent()) {
+            MentorshipRequest request = requestOptionalRequest.get();
```
Как я понял, вот этим if (requestOptionalRequest.isPresent()) ты попытался найти сам Request. А что если не делать if а просто еще при создании ...mentorshipRequestRepository.findById(id) добавить в конце .orElseThrow(()... c выбросом исключения

SlavaVermillion commented on src/main/java/school/faang/user_service/service/MentorshipRequestService.java:
```diff
@@ -0,0 +1,106 @@
+            throw new IllegalArgumentException("there is no such id");
+        }
+
+    }
```
92 строчку можешь убрать, она лишняя)
Ну и вот это else {
            throw new IllegalArgumentException("there is no such id");
        }
уже вряд ли пригодиться

SlavaVermillion replied:
Ну и можно еще явно сохранить requestOptionalRequest

SlavaVermillion commented on src/main/java/school/faang/user_service/service/MentorshipRequestService.java:
```diff
@@ -0,0 +1,106 @@
+@Service
+@RequiredArgsConstructor
+public class MentorshipRequestService {
+    private final MentorshipRequestRepository mentorshipRequestRepository;
```
добавь еще @Transactional для того чтобы при завершении без ошибок все изменения зафиксировались, а при получении ошибок все откатилось  назад и ничего не сохранилось

SlavaVermillion commented on src/main/java/school/faang/user_service/service/MentorshipRequestService.java:
```diff
@@ -0,0 +1,106 @@
+
+    public void rejectRequest(Long id, RejectionDto rejection) {
+        Optional<MentorshipRequest> requestOptionalRequest = mentorshipRequestRepository.findById(id);
+        if (requestOptionalRequest.isPresent()) {
```
Давай вот тут Optional<MentorshipRequest> requestOptionalRequest = mentorshipRequestRepository.findById(id) сразу сделаем .or.ElseThrow(()... исключение

SlavaVermillion commented on src/main/java/school/faang/user_service/service/MentorshipRequestService.java:
```diff
@@ -0,0 +1,106 @@
+
+    }
+
+    public void rejectRequest(Long id, RejectionDto rejection) {
```
Давай добавим @Transactional для сохранения изменений

SlavaVermillion commented on src/main/java/school/faang/user_service/service/MentorshipRequestService.java:
```diff
@@ -0,0 +1,106 @@
+        return requestStream;
+    }
+
+    public void acceptRequest(Long id) {
```
Давай вот сюда @Transactional добавим для сохранения

SlavaVermillion commented on src/main/java/school/faang/user_service/service/MentorshipRequestService.java:
```diff
@@ -0,0 +1,106 @@
+            MentorshipRequest request = requestOptionalRequest.get();
+            request.setStatus(RequestStatus.REJECTED);
+            request.setRejectionReason(rejection.reason());
+        } else {
```
Еще бы желательно добавить проверку на статус запроса, чтобы случайно не отклонить запрос два раза

SlavaVermillion commented:
Хорошее решение, устранить мелкие недочеты и будет супер!

VladimirEkgardt committed:
Исправил замечания Славы. Добавил фильтр DescriptionFilter, который обрабатывает есть ли слово в тексте запроса

VladimirEkgardt committed:
Испраивл запросы, теперь все работает, все запросы отправляются, в бд отображаются

VladimirEkgardt committed:
Добавил проверку в методы rejectRequest и acceptRequest класса MentorshipRequestService

SlavaVermillion reviewed (approved):
Замечания исправлены, супер!

SnurnicynKonstantin commented on src/main/java/school/faang/user_service/controller/MentorshipRequestController.java:
```diff
@@ -0,0 +1,47 @@
+
+    @PostMapping
+    public MentorshipRequestDto requestMentorship(@RequestBody MentorshipRequestDto request) {
+        if (!request.getDescription().isEmpty()) {
```
if отсюда надо дуалить.
Вместо этого на поле description надо повешать аннотацию @notEmpty. А здесь перед RequestBody добавить @valid.

SnurnicynKonstantin commented on src/main/java/school/faang/user_service/dto/MentorshipRequestDto.java:
```diff
@@ -0,0 +1,17 @@
+import java.time.LocalDateTime;
+
+@Data
+public class MentorshipRequestDto {
```
Можно ли переделать на record?

SnurnicynKonstantin commented on src/main/java/school/faang/user_service/dto/RequestFilterDto.java:
```diff
@@ -0,0 +1,13 @@
+import school.faang.user_service.entity.RequestStatus;
+
+@Data
+public class RequestFilterDto {
```
Можно ли на record переделать?

SnurnicynKonstantin commented on src/main/java/school/faang/user_service/repository/mentorship/MentorshipRequestRepository.java:
```diff
@@ -1,18 +1,19 @@
             VALUES (?1, ?2, ?3, 0, NOW(), NOW())
             """)
-    MentorshipRequest create(long requesterId, long receiverId, String description);
+    void create(long requesterId, long receiverId, String description);
```
Тут kexit сделать следующее. `returning *` надо добавить, еще и `@Modifying `удалить. На сколько помню `Modifying` здесь вызовет ошибку. Но могу ошибаться.
Но что бы MentorshipRequest возвращать просто измени sql запрос:
```
INSERT INTO mentorship_request (requester_id, receiver_id, description, status, created_at, updated_at)
VALUES (?1, ?2, ?3, 0, NOW(), NOW())
returning *
```

VladimirEkgardt replied:
А для чего тут возращать что-то?

VladimirEkgardt replied:
Да, все верно так бы Modifying ругался бы, говоря что хочет либо void, либо int/Integer

SnurnicynKonstantin commented on src/main/java/school/faang/user_service/service/filters/RequesterFilter.java:
```diff
@@ -0,0 +1,21 @@
+import java.util.stream.Stream;
+
+@Component
+public class RequesterFilter implements RequestFilter {
```
Вынеси папку filter на общий уровень с controller, service, ... И создай там папку mentorshp. И уже в mentorship положи все фильтры из этой задачки. 
Так мы не спутаем твои фильтры с чужими. Коллеги то же фиотры делают

SnurnicynKonstantin commented on src/main/java/school/faang/user_service/service/MentorshipRequestService.java:
```diff
@@ -0,0 +1,125 @@
+            mentorshipRequestRepository
+                    .create(request.getRequesterId(), request.getReceiverId(), request.getDescription());
+        }
+        return request;
```
Если вернешь request, то там может быть не актуальная инфа из БД. Как минимум id. Лучше маппером entity преобразуй в дто. Если внести правки, которые я рекомендовал по репозиторию, то метод mentorshipRequestRepository.create() должен вернуть объект. Вот его и верни (дто) в контроллере.

SnurnicynKonstantin commented on src/main/java/school/faang/user_service/service/MentorshipRequestService.java:
```diff
@@ -0,0 +1,125 @@
+        MentorshipRequest request = mentorshipRequestRepository.findById(id)
+                .orElseThrow(() -> new IllegalArgumentException("there is no such id"));
+
+        User requester = userRepo.findById(request.getRequester().getId())
```
Этот запрос и запрос ниже повторяются в предыдущем методе.
Лучше создай UserService. В нем метод getUserById. В логике этого метода так же пропиши 
```
userRepo.findById(userId)
                .orElseThrow(() -> new IllegalArgumentException
                        ("There is not user with id =" + userId));
```
А здесь просто вызывай метод у UserService и получай юзера

VladimirEkgardt committed:
Переместил пакет с фильтрами в другой пакет

SnurnicynKonstantin commented on src/main/java/school/faang/user_service/service/MentorshipRequestService.java:
```diff
@@ -0,0 +1,125 @@
+            request.setStatus(RequestStatus.ACCEPTED);
+            mentorshipRequestRepository.save(request);
+        } else {
+            throw new IllegalArgumentException("You already have such a mentor.");
```
Лучше мествами поменять. 
```
if (receiver.getMentees().contains(requester)) {
    throw new IllegalArgumentException("You already have such a mentor.");
} else {
...
}
```
Так легче читать

SnurnicynKonstantin commented on src/main/java/school/faang/user_service/service/MentorshipRequestService.java:
```diff
@@ -0,0 +1,125 @@
+            receiver.getMentees().add(requester);
+            requester.getMentors().add(receiver);
+
+            if (!request.getStatus().equals(RequestStatus.PENDING)) {
```
"The request has already been rejected" - не факт, он может и в статусе accepted быть.
По задачке не требуют таких проверок. А значит статусы могут меняться как угодно. Не нужно тогда таких ограничений накладывать

SnurnicynKonstantin commented on src/main/java/school/faang/user_service/service/MentorshipRequestService.java:
```diff
@@ -0,0 +1,125 @@
+
+    @Transactional
+    public void rejectRequest(Long id, RejectionDto rejection) {
+        MentorshipRequest request = mentorshipRequestRepository.findById(id)
```
Операция то же повторяется. Вынеси ее лучше в отдельный метод. И используй здесь этот метод

SnurnicynKonstantin commented on src/main/java/school/faang/user_service/dto/MentorshipRequestDto.java:
```diff
@@ -0,0 +1,17 @@
+
+@Data
+public class MentorshipRequestDto {
+    private Long id;
```
В запросе не нужен id. Мы же его не знаем еще, мы только создаем запрос

VladimirEkgardt replied:
А на выходе он нам не нужен будет?

SnurnicynKonstantin replied:
Нужен будет. Для запроса ты используешь MentorshipRequestDto. Это понятно по слову Request в названии. В запросе не нужен id.
Но для ответа надо использовать MentorshipResponseDto. Т.е. другое дто. И в нем уже должен быть id

SnurnicynKonstantin commented on src/main/java/school/faang/user_service/controller/MentorshipRequestController.java:
```diff
@@ -0,0 +1,47 @@
+    private final MentorshipRequestService service;
+
+    @PostMapping
+    public MentorshipRequestDto requestMentorship(@RequestBody MentorshipRequestDto request) {
```
Здесь в ответе надо MentorshipResponseDto использовать. Request для запросов только

SnurnicynKonstantin commented on src/main/java/school/faang/user_service/service/MentorshipRequestService.java:
```diff
@@ -0,0 +1,125 @@
+                .findLatestRequest(request.getRequesterId(), request.getReceiverId());
+        if (optionalMentorshipRequest.isPresent()) {
+            MentorshipRequest mentorshipRequest = optionalMentorshipRequest.get();
+            if (ChronoUnit.MONTHS.between(mentorshipRequest.getUpdatedAt(), LocalDate.now()) > 3) {
```
На этой строчке падает ошибка: Unable to obtain LocalDateTime from TemporalAccessor: 2025-05-13 of type java.time.LocalDate
Надо поправить

VladimirEkgardt committed:
Исправил замечания

SnurnicynKonstantin commented on src/main/java/school/faang/user_service/service/MentorshipRequestService.java:
```diff
@@ -0,0 +1,113 @@
+        Optional<MentorshipRequest> optionalMentorshipRequest = mentorshipRequestRepository
+                .findLatestRequest(request.requesterId(), request.receiverId());
+
+        MentorshipRequest mentorshipRequestReturn;
```
Лучше переименовать в newMentorshipRequest. Слово Return не понятно

SnurnicynKonstantin commented on src/main/java/school/faang/user_service/filters/mentorship_request/DescriptionFilter.java:
```diff
@@ -0,0 +1,20 @@
+package school.faang.user_service.filters.mentorship_request;
```
filters -> filters. Контроллер, сервис, репозиторий, все они в единственном числе. Фильтры то же должны быть в единственном

SnurnicynKonstantin commented on src/main/java/school/faang/user_service/service/MentorshipRequestService.java:
```diff
@@ -0,0 +1,113 @@
+            requester.getSentMentorshipRequests().add(mapper.toEntity(request));
+            receiver.getReceivedMentorshipRequests().add(mapper.toEntity(request));
+
+            mentorshipRequestReturn = mentorshipRequestRepository
```
Операцию 
```
mentorshipRequestReturn = mentorshipRequestRepository
                    .create(request.requesterId(), request.receiverId(), request.description());
```
 И такую же операцию ниже можно вынести из условия. Ты не не зависимо от условия ее выполняешь, по этому после if else можно написать:
```
 MentorshipRequest newMentorshipRequest = mentorshipRequestRepository
                    .create(request.requesterId(), request.receiverId(), request.description());
return mapper.toDto(newMentorshipRequest);
```

VladimirEkgardt committed:
1) В MentorshipRequestDto убрал поле id
2) Создал MentorshipResponseDto
3) MentorshipRequestController
3.1) В меотде requestMentorship изменил тип возращаемого значения на MentorshipResponseDto
3.2) В методе getRequests изменил тип возращаемого значения на List<MentorshipResponseDto>
4) MentorshipRequestService
4.1) В методе requestMentorship переименовал переменную
4.2) Тело проверки поправил
5) Переименовал пакет filters->filter
6) MentorshipRequestMapper добавил mappint(id)
7) Во всех фильтрах поправил баги

SnurnicynKonstantin commented on src/main/java/school/faang/user_service/service/MentorshipRequestService.java:
```diff
@@ -0,0 +1,126 @@
+        receiver.getReceivedMentorshipRequests().add(mapper.toEntity(request));
+        //Будут ли изменения выше сохраняться в базе?
+
+        newMentorshipRequest = mentorshipRequestRepository
```
Зачем переменную там создавать, а инициализировать тут?
Можно было же сразу на 60й строчке создать и инициализировать. Я в комменте своем писал про это
<img width="949" alt="image" src="https://github.com/user-attachments/assets/74714ad2-6e69-4613-b736-13fa9f878009" />

SnurnicynKonstantin commented on src/main/java/school/faang/user_service/service/MentorshipRequestService.java:
```diff
@@ -0,0 +1,126 @@
+        return mentorshipRequestRepository.findById(id).orElseThrow(() -> new IllegalArgumentException("there is no such id"));
+    }
+
+    private MentorshipResponseDto toMentorshipResponseDto(MentorshipRequest request) {
```
Для таких вещей маппер используем

VladimirEkgardt replied:
Я тоже подумал об этом, но не был уверен, что делают мапперы для двух дто

VladimirEkgardt committed:
Добавил RequestToResponseDto mapper

SnurnicynKonstantin commented on src/main/java/school/faang/user_service/service/MentorshipRequestService.java:
```diff
@@ -0,0 +1,115 @@
+
+        requester.getSentMentorshipRequests().add(requestMapper.toEntity(request));
+        receiver.getReceivedMentorshipRequests().add(requestMapper.toEntity(request));
+        //Будут ли изменения выше сохраняться в базе?
```
Запрос mentorshipRequestRepository.create() и так свяжет юзера и запрос. Так что можно не делать две операции выше. Они никакой роли не играют

SnurnicynKonstantin reviewed (approved):
Один коммент оставил. ПР апрувнул. Давай удалим две лишние строки и будем мержить. 
Как удалишь, на ревью переводи опять

VladimirEkgardt committed:
Исправил замечания, нашел пару багов

VladimirEkgardt committed:
Переименовал RequestFilterDto на MentorshipRequestFilterDto, создал пакет mentorship_request в пакете dto, переместил все dto в этот паакет
